### PR TITLE
Freeport 2.0

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

New freeport implementation; taken from battle-tested Temporal projects.

## Why?

Follow-up to https://github.com/temporalio/temporal/pull/6915; making it simpler and more robust.

## How did you test it?

- [x] works on Linux (ie CI)
- [x] works on Mac (ie my computer)

[First CI run](https://github.com/temporalio/temporal/actions/runs/12269035797/attempts/1?pr=6966), `TestNexusCallbackReplicated` failed with `Error while dialing: dial tcp 127.0.0.1:33279: connect: connection refused`. Looks like there was an unrelated data race (that I've reported to Slack).

[Second CI run](https://github.com/temporalio/temporal/actions/runs/12269035797/attempts/2?pr=6966) passed.

[Third CI run](https://github.com/temporalio/temporal/actions/runs/12269035797?pr=6966), Versioning suite seems to have [timed out](https://github.com/temporalio/temporal/actions/runs/12269035797/job/34233452135?pr=6966#step:8:966). Unrelated.

So it seems to be fine from a random port perspective.

### CODE REVIEW

File: internal/freeport/freeport.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// The MIT License
```
Comment (stephanos) [inline internal/freeport/freeport.go:1]: There are 2 copies of this across Temporal:
- https://github.com/temporalio/cli/blob/main/temporalcli/devserver/freeport.go
- https://github.com/temporalio/sdk-go/blob/master/testsuite/freeport.go

File: internal/freeport/freeport.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// within a few seconds.
+//
+// On Unix-based systems, binding to the port returned by this function requires
+// setting the `SO_REUSEADDR` socket option (Go already does that by default,
+// but other languages may not); otherwise, the OS may fail with a message such
+// as "address already in use". Windows default behavior is already appropriate
+// in this regard; on that platform, `SO_REUSEADDR` has a different meaning and
+// should not be set (setting it may have unpredictable consequences).
+func MustGetFreePort() int {
+	port, err := getFreePort("127.0.0.1")
```
Comment (bergundy) [inline internal/freeport/freeport.go:47]: What about ipv6? The temporalite version of this tried to listen on a `tcp6` address.

File: internal/freeport/freeport.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
```
Comment (bergundy) [inline internal/freeport/freeport.go:3]:
```suggestion
// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
```

**Review by bergundy — APPROVED** [review]
Thanks!

File: internal/freeport/freeport.go (hunk 3 above)
Comment (stephanos) [reply internal/freeport/freeport.go:3]: I _believe_ the server license lint check requires 2022. Let me double-check.

Comment (stephanos) [reply internal/freeport/freeport.go]: Huh, I must have imagined that; fixed!

File: internal/freeport/freeport.go (hunk 2 above)
Comment (stephanos) [reply internal/freeport/freeport.go:47]: 👍 The CLI doesn't have that, so I thought it was fine. But I can see that the SDK impl has it, so I copied that part over now, too.

**Pull request merged**